### PR TITLE
Put an abstract type over the optimisers to allow dispatch

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Required for https://github.com/SciML/GalacticOptim.jl/pull/255

### CODE REVIEW

**Review by DhairyaLGandhi — APPROVED** [review]
Lgtm

**Comment by mloubout** [comment]
Should the abstract type be parametric for safety?

**Comment by ChrisRackauckas** [comment]
Making that work for the OptimiserChain is difficult so I skipped that for now, but that could always be added with backwards compatibility (since the no parameter form would still work the same).

File: src/rules.jl [hunk 1]
```diff
@@ -6,6 +6,8 @@
 @deprecate ADAGrad AdaGrad
 @deprecate ADADelta AdaDelta
 
+abstract type AbstractOptimiser end
```
Comment (mcabbott) [inline src/rules.jl:9]: If we do this, the name should be something like AbstractRule

**Review by mcabbott — CHANGES_REQUESTED** [review]
This needs some thought, and approval by someone involved in the package. 

Can motivate a little why this is a good idea, for those not following GalacticOptim closely? Why change only this file, not the others, not the docs?

**Comment by ChrisRackauckas** [comment]
> Can motivate a little why this is a good idea, for those not following GalacticOptim closely?

We need to dispatch for this package's optimizers. The only other way to do that is to maintain a `Union` of all optimizers, which is a nasty kludge.

**Comment by darsnack** [comment]
Unfortunately, we have the same super type in Flux for distinguishing between the old Flux optimisers and this package. I suppose we can always explicitly name `Flux.AbstractOptimiser`, etc. since only FluxML packages will try to make this distinction. User code shouldn't be using the dispatch.

File: src/rules.jl [hunk 2]
```diff
@@ -16,7 +16,7 @@ For each parameter `p` and its gradient `dp`, this runs `p -= η*dp`.
 - Learning rate (`η`): Amount by which gradients are discounted before updating
                        the weights.
 """
-struct Descent{T}
+struct Descent{T} <: Rule
   eta::T
 end
```
[the remark this answers is not in the document]
Comment (mcabbott) [reply src/rules.jl]: Now `AbstractRule` as that seems clearer

**Review by darsnack — APPROVED** [review]

**Comment by darsnack** [comment]
On second thought, does restricting the types to `<:Real` constitute a breaking change? In some sense, it probably won't work with other types, so it could be a bug fix. But it will affect serialization of the optimizers with BSON, and I think that's been one use case for preferring Optimisers.jl over Flux's optimizers.

Can we separate these changes out (only do `AbstractRule` here)?

**Comment by mloubout** [comment]
What's the motivation behind the `<:Real` restriction?  One may wanna use complex-valued optimizers for some problems such as PINN.

**Comment by mloubout** [comment]
I mean complex-valued learning rate. I don't have a reference but this is something that may be of interest even though probably very rare and exoctic.

**Comment by mcabbott** [comment]
I can't picture a complex learning rate in an existing optimiser being anything but a mistake, and I'd prefer mistakes to be give errors as soon as possible. If you define your own, then it's clearer that you own the consequences & should test it, rather than hoping that this package has thought about it already.

At present you could also supply say a matrix, but the package isn't in fact generic enough to allow that... and probably shouldn't be, without a real need and a lot of tests: 
```julia
julia> st = Optimisers.setup(Descent([1 2; 3 4]), [[5 6; 7 8.0]])
1-element Vector{Optimisers.Leaf{Descent{Matrix{Int64}}, Nothing}}:
 Leaf(Descent{Matrix{Int64}}([1 2; 3 4]), nothing)

julia> Optimisers.update!(st, [[5 6; 7 8]], [ones(2,2)])
ERROR: MethodError: Cannot `convert` an object of type Matrix{Int64} to an object of type Float64
Closest candidates are:
  convert(::Type{T}, ::T) where T<:Number at number.jl:6
  convert(::Type{T}, ::Number) where T<:Number at number.jl:7
  convert(::Type{T}, ::Base.TwicePrecision) where T<:Number at twiceprecision.jl:273
  ...
Stacktrace:
 [1] apply!(o::Descent{Matrix{Int64}}, state::Nothing, x::Matrix{Int64}, dx::Matrix{Float64})
   @ Optimisers ~/.julia/packages/Optimisers/FQWDl/src/rules.jl:27
 [2] update!(ℓ::Optimisers.Leaf{Descent{Matrix{Int64}}, Nothing}, x::Matrix{Int64}, x̄s::Matrix{Float64})
   @ Optimisers ~/.julia/packages/Optimisers/FQWDl/src/interface.jl:31
```
I agree that before 0.2 would have been the better time to restrict such types. Would be easy to revert.

**Comment by mcabbott** [comment]
> But it will affect serialization of the optimizers with BSON

Can you say more? The issue is loading things saved before such a change once you have new struct definitions? And is it sensitive to type restrictions but not to subtyping?

**Comment by mloubout** [comment]
>I can't picture a complex learning rate in an existing optimiser being anything but a mistake

How so? There is work on complex-valued optimizers (i.e [this](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=7321072) and being able to use the optimizer out of the box makes it a lot easier to check these. Takes away a bit from using this package as well if any slight changes requires to make a new optimizer but that's just me.

**Comment by ToucheSir** [comment]
> Can you say more? The issue is loading things saved before such a change once you have new struct definitions? And is it sensitive to type restrictions but not to subtyping?

As long as people weren't saving their optimizers with non-real hyperparams, adding the type constraint should be a non-issue.

> Takes away a bit from using this package as well if any slight changes requires to make a new optimizer but that's just me.

And are you doing this? I don't think it would be the end of the world to relax the constraint to `<:Number`, but my reading of the discussion so far is that the idea of using non-real hyperparams for existing first-order optimizers used in ML is a purely theoretical one. Developing new optimizers yes, but do we know if e.g. RMSProp will still function at all if given a complex-valued learning rate with both components having opposite signs?

**Comment by mcabbott** [comment]
Ok, I am surprised by this complex-η paper, thanks for the link.

> e.g. RMSProp will still function

Yes this is my concern. It might be that there's a right way to extend it, which involves conjugating some of the terms. But changing that once people rely on the earlier naiive (untested) behaviour becomes difficult. Surely the RMSProp paper we follow did not have anything but small positive η in mind.

**Comment by darsnack** [comment]
> Can you say more? The issue is loading things saved before such a change once you have new struct definitions? And is it sensitive to type restrictions but not to subtyping?

Specifically, the issue is whether the new constructor will be able to accept what was saved when loading with a new definition. Subtyping won't affect it. Admittedly, as Brian points out, the chances are low that someone runs into this.

**Comment by darsnack** [comment]
In any case, it seems like there are two things in this PR. One is ready to go and the other seems to still have discussion. Let's just separate the two since it seems like SciML is wanting one now but not the other.

**Review by darsnack — APPROVED** [review]

**Pull request merged**